ext/opentelemetry-ext-system-metrics/src/opentelemetry/ext/system_metrics/__init__.py

```diff
@@ -0,0 +1,182 @@
+from opentelemetry.sdk.metrics.export.controller import PushController
+
+_DEFAULT_INTERVAL = 30
+
+
+class SystemMetrics:
+    def __init__(
+        self,
+        exporter: MetricsExporter,
+        interval: int = _DEFAULT_INTERVAL,
+        labels: typing.Optional[typing.Dict[str, str]] = None,
+    ):
+        labels = {} if labels is None else labels
+        self.meter = metrics.get_meter(__name__)
+        self.controller = PushController(
+            meter=self.meter, exporter=exporter, interval=interval
+        )
+        self._proc = psutil.Process(os.getpid())
+        self._labels = labels
+        self._system_memory_labels = {}
+        self._system_cpu_labels = {}
+        self._network_bytes_labels = {}
+        self._runtime_memory_labels = {}
+        self._runtime_gc_labels = {}
+        # create the label set for each observer once
+        for key, value in self._labels.items():
+            self._system_memory_labels[key] = value
+            self._system_cpu_labels[key] = value
+            self._network_bytes_labels[key] = value
+            self._runtime_memory_labels[key] = value
+            self._runtime_gc_labels[key] = value
+
+        self.meter.register_observer(
+            callback=self._get_system_memory,
+            name="system.mem",
+            description="System memory",
+            unit="bytes",
+            value_type=int,
+            label_keys=self._labels.keys(),
+        )
+
+        self.meter.register_observer(
+            callback=self._get_system_cpu,
+            name="system.cpu",
+            description="System CPU",
+            unit="seconds",
+            value_type=float,
+            label_keys=self._labels.keys(),
+        )
+
+        self.meter.register_observer(
+            callback=self._get_network_bytes,
+            name="system.net.bytes",
+            description="System network bytes",
+            unit="bytes",
+            value_type=int,
+            label_keys=self._labels.keys(),
+        )
+
+        self.meter.register_observer(
+            callback=self._get_runtime_memory,
+            name="runtime.python.mem",
+            description="Runtime memory",
+            unit="bytes",
+            value_type=int,
+            label_keys=self._labels.keys(),
+        )
+
+        self.meter.register_observer(
+            callback=self._get_runtime_gc_count,
+            name="runtime.python.gc.count",
+            description="Runtime: gc objects",
+            unit="objects",
+            value_type=int,
+            label_keys=self._labels.keys(),
+        )
+
+    def _get_system_memory(self, observer: metrics.Observer) -> None:
+        """Observer callback for memory available
+
+        Args:
+            observer: the observer to update
+        """
+        system_memory = psutil.virtual_memory()
+        for key, value in system_memory._asdict().items():
+            if not isinstance(value, int):
```

Comment:
This looks like some trouble. We are using a private attribute here and the criterion we are using to select which memory metrics are to be observed is that the type of the value is `int`. I checked the attributes of a `svmem` object returned by `psutil.virtual_memory` and only `percent` (which seems to be `100 * (1 - available / total)`) is not of type `int`. Nevertheless, I wonder if we actually want to include all the rest of `int`-typed attributes to be observed or what if in the future an interesting non-`int` attribute shows up. Maybe we should pass to the observer a fixed set of the attributes of the object returned by `psutil.virtual_memory` instead?

Reply:
After reviewing your comment, I decided as a first iteration to just collect the metrics that are the same on all the platforms as per the psutil documentation.